cmd/gameserver/packet/inventory.go

```diff
@@ -0,0 +1,130 @@
+	send := character.ItemMoveReq{Server: byte(g_ServerSettings.ServerId), Id: id, DeleteSlot: deleteSlot, CreateSlot: createSlot}
+	receive := character.ItemMoveRes{}
+	err := g_RPCHandler.Call(packet, send, &receive)
+	if err != nil {
+		log.Error(err.Error())
+	}
+
+	return receive
+}
+
+func StorageExchangeMove(session *network.Session, reader *network.Reader) {
+	equip := reader.ReadUint32()
+	deleteSlot := uint16(reader.ReadUint32())
+	inventory := reader.ReadUint32()
+	createSlot := uint16(reader.ReadUint32())
+
+	ctx, err := parseSessionContext(session)
+	if err != nil {
+		log.Error(err.Error())
+		SendStorageExchangeMoveStatusPacket(session, 0x00)
+		return
+	}
+	playerChar := ctx.char
+
+	isEquip := equip == 1
+	isInventory := inventory == 1
+
+	if isEquip && !isInventory {
+		result := SendStorageExchangeMoveMasterServerPacket(rpc.UnEquipItem, playerChar.Id, deleteSlot, createSlot)
+
+		if result.Result == nil {
+			playerChar.Equipment.Remove(deleteSlot)
+			//playerChar.Inventory.Set(createSlot, result.Item)
+
+			pkt := network.NewWriter(NFY_ITEM_UNEQUIP)
+			pkt.WriteInt32(playerChar.Id)
+			pkt.WriteInt16(deleteSlot)
+
+			g_NetworkManager.SendToAll(pkt)
+
+			SendStorageExchangeMoveStatusPacket(session, 0x01)
+			return
+		} else {
+			log.Error(result.Result)
+			SendStorageExchangeMoveStatusPacket(session, 0x00)
+			return
+		}
+	}
+
+	if isInventory && !isEquip {
+		result := SendStorageExchangeMoveMasterServerPacket(rpc.EquipItem, playerChar.Id, deleteSlot, createSlot)
+
+		if result.Result == nil {
+			playerChar.Inventory.Remove(deleteSlot)
+			playerChar.Equipment.Set(createSlot, result.Item)
+
+			pkt := network.NewWriter(NFY_ITEM_EQUIP)
+			pkt.WriteInt32(playerChar.Id)
+			pkt.WriteInt32(result.Item.Kind)
+			pkt.WriteInt16(result.Item.Slot)
+			pkt.WriteByte(0x00)
+			pkt.WriteByte(0x00)
+			pkt.WriteByte(0x00)
+			pkt.WriteByte(0x00)
+			pkt.WriteByte(0x00)
+
+			g_NetworkManager.SendToAll(pkt)
+
+			SendStorageExchangeMoveStatusPacket(session, 0x01)
+			return
+		} else {
+			log.Error(result.Result)
+			SendStorageExchangeMoveStatusPacket(session, 0x00)
+			return
+		}
+	}
+
+	if isEquip && isInventory {
+		result := SendStorageExchangeMoveMasterServerPacket(rpc.ChangeEquipItemSlot, playerChar.Id, deleteSlot, createSlot)
+		if result.Result == nil {
+			playerChar.Equipment.Remove(deleteSlot)
+			playerChar.Equipment.Set(createSlot, result.Item)
+
+			//TODO Check and add notify packets
+
+			SendStorageExchangeMoveStatusPacket(session, 0x01)
+			return
+		} else {
+			log.Error(result.Result)
+			SendStorageExchangeMoveStatusPacket(session, 0x00)
+			return
+		}
+	}
+
+	if !isEquip && !isInventory {
+		result := SendStorageExchangeMoveMasterServerPacket(rpc.ChangeInventoryItemSlot, playerChar.Id, uint16(deleteSlot), uint16(createSlot))
+
+		if result.Result == nil {
+			playerChar.Inventory.Remove(deleteSlot)
+			//playerChar.Inventory.Set(createSlot, result.Item)
```

Comment:
I redid it, please look, and for some reason it does not work

`playerChar.Inventory.Set(createSlot, result.Item)`

I couldn't figure it out

Reply:
You mean it doesn't add into inventory map? It's been years since I've wrote implementation for it and I never properly tested it, I will take a look into this.

Reply:
It throws an error
![image](https://github.com/ubis/Freya/assets/34573976/23653650-d6ed-45f2-80d7-143d65280380)

I corrected some things, but I'm not sure if I did everything right

Reply:
Ah, it's because inventory is not assigned(initialized) in character struct.

It should work if you would add:

```Go
c.Inventory = res.Inventory
```

in https://github.com/ubis/Freya/blob/feature/inventory_move/cmd/gameserver/packet/initialized.go#L211

Can you check this?

Reply:
Yes you are right, why didn't I figure it out myself, thanks I added it to the this PR

Reply:
Does it work now? I think we can merge this

Reply:
Yep i'ts work now